Library/OcAppleKernelLib/MkextContext.c

```diff
@@ -0,0 +1,1040 @@
+      return EFI_BUFFER_TOO_SMALL;
+    }
+
+    //
+    // Plist will be placed at end of mkext.
+    //
+    PlistOffset = Context->MkextSize;
+    if (OcOverflowAddU32 (PlistOffset, MKEXT_ALIGN (InfoPlistSize), &MkextNewSize)) {
+      return EFI_INVALID_PARAMETER;
+    } else if (MkextNewSize > Context->MkextAllocSize) {
+      return EFI_BUFFER_TOO_SMALL;
+    }
+
+    //
+    // Executable, if present, will be placed right after plist.
+    //
+    if (Executable != NULL) {
+      ASSERT (ExecutableSize > 0);
+
+      BinOffset = MkextNewSize;
+      if (!ParseKextBinary (&Executable, &ExecutableSize, Context->Is64Bit)
+        || OcOverflowAddU32 (BinOffset, MKEXT_ALIGN (ExecutableSize), &MkextNewSize)) {
+        return EFI_INVALID_PARAMETER;
+      }
+      
+      if (MkextNewSize > Context->MkextAllocSize) {
+        return EFI_BUFFER_TOO_SMALL;
+      }
+
+      CopyMem (&Context->Mkext[BinOffset], Executable, ExecutableSize);
+      Context->MkextHeader->V1.Kexts[Context->NumKexts].Binary.Offset = SwapBytes32 (BinOffset);
+      Context->MkextHeader->V1.Kexts[Context->NumKexts].Binary.CompressedSize = 0;
+      Context->MkextHeader->V1.Kexts[Context->NumKexts].Binary.FullSize = SwapBytes32 (ExecutableSize);
+      Context->MkextHeader->V1.Kexts[Context->NumKexts].Binary.ModifiedSeconds = 0;
+    }
+
+    CopyMem (&Context->Mkext[PlistOffset], InfoPlist, InfoPlistSize);
+    Context->MkextHeader->V1.Kexts[Context->NumKexts].Plist.Offset = SwapBytes32 (PlistOffset);
+    Context->MkextHeader->V1.Kexts[Context->NumKexts].Plist.CompressedSize = 0;
+    Context->MkextHeader->V1.Kexts[Context->NumKexts].Plist.FullSize = SwapBytes32 (InfoPlistSize);
+    Context->MkextHeader->V1.Kexts[Context->NumKexts].Plist.ModifiedSeconds = 0;
+
+    //
+    // Assumption:
+    //    NumKexts is checked to be under MaxNumKexts, which is assumed to be under
+    //    the max value of UINT32 during context creation. This operation thus cannot overflow.
+    //
+    Context->MkextSize = MkextNewSize;
+    Context->NumKexts++;
+
+  //
+  // Mkext v2.
+  //
+  } else if (Context->MkextVersion == MKEXT_VERSION_V2) {
+    PlistBuffer = AllocateCopyPool (InfoPlistSize, InfoPlist);
+    if (PlistBuffer == NULL) {
+      return EFI_OUT_OF_RESOURCES;
+    }
+
+    PlistXml = XmlDocumentParse (PlistBuffer, InfoPlistSize, FALSE);
+    if (PlistXml == NULL) {
+      FreePool (PlistBuffer);
+      return EFI_INVALID_PARAMETER;
+    }
+
+    PlistRoot = PlistNodeCast (PlistDocumentRoot (PlistXml), PLIST_NODE_TYPE_DICT);
+    if (PlistRoot == NULL) {
+      XmlDocumentFree (PlistXml);
+      FreePool (PlistBuffer);
+      return EFI_INVALID_PARAMETER;
+    }
+
+    //
+    // We are not supposed to check for this, it is XNU responsibility, which reliably panics.
+    // However, to avoid certain users making this kind of mistake, we still provide some
+    // code in debug mode to diagnose it.
+    //
+    DEBUG_CODE_BEGIN ();
```

Comment:
Make this a function, as it really is copied from PK code. In fact, maybe starting with `PlistBuffer` assignment.